src/eval/run/system.rs

```diff
@@ -79,15 +87,31 @@ pub fn create<M: Memory + Default, P: Patch>(state: &mut State<M, P>, after_gas:
     try_balance!(state, value, Gas::zero());
 
     let init = Rc::new(copy_from_memory(&state.memory, init_start, init_len));
-    let transaction = ValidTransaction {
-        caller: Some(state.context.address),
-        gas_price: state.context.gas_price,
-        gas_limit: l64_after_gas,
-        value,
-        input: init,
-        action: TransactionAction::Create,
-        nonce: state.account_state.nonce(state.context.address).unwrap(),
+    let transaction = if !is_create2 {
+        ValidTransaction {
+            caller: Some(state.context.address),
+            gas_price: state.context.gas_price,
+            gas_limit: l64_after_gas,
+            value,
+            input: init,
+            action: TransactionAction::Create,
+            nonce: state.account_state.nonce(state.context.address).unwrap(),
+        }
+    } else {
+        pop!(state, salt: H256);
+        let init_hash = Keccak256::digest(&init);
```

Comment:
Better to provide different names, or chain calls if intermediate values unnecessary as variables.

Reply:
It's a common practice in rust to shadow intermediate values if they replace each other and not used separately. Makes no sense to me to make a mess with chained calls here:

```
let init_hash = Keccak256::digest(&init);
let init_hash = M256::from(&init_hash[..]);
```
is better then
```
let init_hash = M256::from(&Keccak256::digest(&init)[..]);
```
The second one is too messy imo.

Reply:
ok, lets leave it